homeassistant/components/smartthings/climate.py

```diff
@@ -0,0 +1,226 @@
+        else:
+            heating_setpoint = kwargs.get(ATTR_TARGET_TEMP_LOW)
+            cooling_setpoint = kwargs.get(ATTR_TARGET_TEMP_HIGH)
+        tasks = []
+        if heating_setpoint is not None:
+            tasks.append(self._device.set_heating_setpoint(
+                round(heating_setpoint, 3), set_status=True))
+        if cooling_setpoint is not None:
+            tasks.append(self._device.set_cooling_setpoint(
+                round(cooling_setpoint, 3), set_status=True))
+        await asyncio.gather(*tasks)
+
+        # State is set optimistically in the commands above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_update(self):
+        """Update the device."""
+        # Flags need to be set dynamically based on operation mode.
+        flags = self._base_supported_features
+        if self.current_operation == STATE_AUTO:
+            flags |= SUPPORT_TARGET_TEMPERATURE_LOW \
+                     | SUPPORT_TARGET_TEMPERATURE_HIGH
+        elif self.current_operation in (STATE_HEAT, STATE_COOL):
+            flags |= SUPPORT_TARGET_TEMPERATURE
+        self._supported_features = flags
+
+    @property
+    def current_fan_mode(self):
+        """Return the fan setting."""
+        return self._device.status.thermostat_fan_mode
+
+    @property
+    def current_humidity(self):
+        """Return the current humidity."""
+        return self._device.status.humidity
+
+    @property
+    def current_operation(self):
+        """Return current operation ie. heat, cool, idle."""
+        return MODE_TO_STATE[self._device.status.thermostat_mode]
+
+    @property
+    def current_temperature(self):
+        """Return the current temperature."""
+        return self._device.status.temperature
+
+    @property
+    def device_state_attributes(self):
+        """Return device specific state attributes."""
+        return {
+            ATTR_OPERATION_STATE:
+                self._device.status.thermostat_operating_state
+        }
+
+    @property
+    def fan_list(self):
+        """Return the list of available fan modes."""
+        return self._device.status.supported_thermostat_fan_modes
+
+    @property
+    def operation_list(self):
+        """Return the list of available operation modes."""
+        return {MODE_TO_STATE[mode] for mode in
+                self._device.status.supported_thermostat_modes}
+
+    @property
+    def supported_features(self):
+        """Return the supported features."""
+        return self._supported_features
+
+    @property
+    def target_temperature(self):
+        """Return the temperature we try to reach."""
+        if self.current_operation == STATE_COOL:
+            return self.target_temperature_high
+        if self.current_operation == STATE_HEAT:
+            return self.target_temperature_low
+        return self.current_temperature
```

Comment:
Why should we return current temperature in this case?

Reply:
More quirks with the entity and this property would go away entirely if the arch issue is resolved.  What is the single-value target temperature when `off` or `eco` or `auto`?  Returning None causes issues in the UI, and again, this is how other platforms solve the issue.

Reply:
If we don't have a single target temperature, we should return `None`. We shouldn't care about UI.

Reply:
Yes, except when I did, it shows a Null in the front, sounds like a bug there?

Reply:
Maybe? Usually we want either single target temp set or both high and low target temp set.

Reply:
Right -- and that depends on what mode we're in.  `heat` has a single temp where as `auto` has both.  The underlying device doesn't see it that way (it just has 2 attributes for heat and cool points, regardless of the mode)--but functionally you can't set both when not in `auto`...  Similar to the comment above about where this domain logic belongs, I think it will have to stay here until I either update the library or create a new library.

Reply:
I suggest looking at the wink climate platform and use similar logic, both here and for set temperature above. Wink takes operation mode in account, and sets the three target temps accordingly, ie single target temp is None if not either heat or cool is active, and low and high target temp are None if not in auto mode.

Reply:
Thanks for the pointer.  The logic was already equivalent with the exception of returning `None`.  I've followed suite to so we can move on.  FYI here's what's displayed in the front-end when `None` is returned.  I don't have a Wink to test, but would expect to see the same thing.  FYI, my Nest thermostats continue to show the heat temperature when off.

![image](https://user-images.githubusercontent.com/6730289/52607991-0ea42800-2e3e-11e9-8097-5f78b43f6dff.png)